LE Audio: Allow server to initiate Codec Config procedure

mkasenberg opened the pull request:
Required by ASCS/SR/ACP/BV-07-C PTS test case.

fixes https://github.com/zephyrproject-rtos/zephyr/issues/46238

mkasenberg committed:
LE Audio: Allow server to initiate Codec Config procedure

Required by ASCS/SR/ACP/BV-07-C PTS test case.

Signed-off-by: Magdalena Kasenberg <magdalena.kasenberg@codecoup.pl>

mkasenberg commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1850,6 +1850,20 @@ int bt_audio_stream_config(struct bt_conn *conn,
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_audio_stream_config_srv(struct bt_conn *conn, int ase_id,
+				struct bt_codec *codec);
```
In the ASCS/SR/ACP/BV-07-C test case the ase_id is given in MMI message, so it can be parsed by AutoPTS and sent to the IUT. But I don't know about other applications.

mkasenberg replied:
I made a refactor and included the `ase_id` parameter in the definition of `bt_audio_stream_config`. Is that ok? Other option I've considered was creating in public api a getter like
`bt_audio_ep *ascs_ep_get(struct bt_conn *conn, uint8_t ase_id);`
and then just pass the retrieved *ep to the function `bt_audio_stream_config`.

mkasenberg commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1842,13 +1842,15 @@ int bt_audio_discover(struct bt_conn *conn,
 			   struct bt_audio_ep *ep,
-			   struct bt_codec *codec);
+			   struct bt_codec *codec,
+			   uint8_t ase_id);
```
I had to use `ase_get`, because `ascs->ases` is empty before initiating Codec Config, but I'm not sure if I can freely use it?

I may have misunderstood how to use `bt_audio_ep_info`, is it ok as it is now?

Thalley commented on subsys/bluetooth/audio/endpoint.h:
```diff
@@ -53,6 +53,10 @@ struct bt_audio_ep {
 	struct bt_audio_broadcast_sink *broadcast_sink;
 };
 
+struct bt_audio_ep_info {
+	struct bt_audio_ep *ep;
+};
```
The struct should contain the value that you want from the `ep`. Since `ep` is an opaque struct, just getting the pointer won't do much good. 

The struct should rather have a field
```c
uint8_t ase_id
```

and assigned like 
```c
ase_id = ep->status.id;
```

We could also add the state (`ep->status.state`) to it, as that has been requested by https://github.com/zephyrproject-rtos/zephyr/issues/46238

Thalley commented on subsys/bluetooth/audio/endpoint.h:
```diff
@@ -134,3 +138,8 @@ static inline const char *bt_audio_ep_state_str(uint8_t state)
 bool bt_audio_ep_is_broadcast_src(const struct bt_audio_ep *ep);
 bool bt_audio_ep_is_unicast_client(const struct bt_audio_ep *ep);
+
+typedef void (*bt_audio_ase_func_t)(const struct bt_audio_ep_info *info,
+		void *user_data);
+void bt_unicast_server_foreach_ase(struct bt_conn *conn,
+		bt_audio_ase_func_t func, void *user_data);
```
These and the `struct bt_audio_ep_info` should be put into the public API in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/bluetooth/audio/audio.h so that it's usable by applications

Thalley commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -349,22 +347,11 @@ int bt_audio_stream_config(struct bt_conn *conn,
 		return -EINVAL;
 	}
 
-	if (stream->conn != NULL) {
-		LOG_DBG("Stream already configured for conn %p", stream->conn);
-		return -EALREADY;
-	}
```
This should be kept

Thalley commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -349,22 +347,11 @@ int bt_audio_stream_config(struct bt_conn *conn,
 	       codec ? codec->vid : 0);
 
-	CHECKIF(conn == NULL || stream == NULL || codec == NULL) {
+	CHECKIF(conn == NULL || codec == NULL || ep == NULL) {
```
We should still have a NULL check for the `stream`

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1377,58 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+	}
+
+	stream = NULL;
+	if (unicast_server_cb != NULL &&
+		unicast_server_cb->config != NULL) {
+		err = unicast_server_cb->config(conn, ep,
+						ep->dir,
+						&ep->codec, &stream,
+						&ep->qos_pref);
+	} else {
+		err = -ENOTSUP;
+	}
```
If the server is configuring a stream via `bt_audio_stream_config` I think we should skip the callback here, and have the server supply the `stream` in the function call instead.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1377,58 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+	 */
+	(void)memcpy(&old_codec, &ep->codec, sizeof(old_codec));
+
+	if (ascs_ep_set_codec(ep, codec->id,
+				sys_le16_to_cpu(codec->cid),
+				sys_le16_to_cpu(codec->vid),
+				NULL, 0, codec)) {
```
```suggestion
	if (ascs_ep_set_codec(ep, codec->id, sys_le16_to_cpu(codec->cid),
			      sys_le16_to_cpu(codec->vid), NULL, 0, codec)) {
```

Indentation fix

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1377,58 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
 	return 0;
 }
 
+int ascs_ase_config(struct bt_conn *conn, struct bt_audio_ep *ep,
+					struct bt_codec *codec)
```
```suggestion
int ascs_ase_config(struct bt_conn *conn, struct bt_audio_ep *ep, 
		    struct bt_codec *codec)
```

Indentation fix

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1377,58 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+	stream = NULL;
+	if (unicast_server_cb != NULL &&
+		unicast_server_cb->config != NULL) {
+		err = unicast_server_cb->config(conn, ep,
+						ep->dir,
+						&ep->codec, &stream,
+						&ep->qos_pref);
```
```suggestion
		err = unicast_server_cb->config(conn, ep, ep->dir, &ep->codec, 
						&stream, &ep->qos_pref);
```

Indentation fix

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1377,58 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+	}
+
+	stream = NULL;
+	if (unicast_server_cb != NULL &&
+		unicast_server_cb->config != NULL) {
```
```suggestion
	if (unicast_server_cb != NULL && unicast_server_cb->config != NULL) {
```

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1377,58 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+
+	ascs_ep_set_state(ep, BT_AUDIO_EP_STATE_CODEC_CONFIGURED);
+
+	if (err != 0) {
+		(void)memcpy(&ep->codec, &old_codec, sizeof(old_codec));
+		LOG_DBG("Failed to configure stream: %d", err);
+
+		return err;
+	}
```
Duplicate of lines 1411-1416?

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1446,6 +1497,31 @@ static ssize_t ascs_config(struct bt_ascs *ascs, struct net_buf_simple *buf)
+	struct bt_ascs_ase *ase;
+	struct bt_ascs *ascs = ascs_get(conn);
+
+	for (size_t i = 1; i <= ASE_COUNT; i++) {
```
Suggest to use 
```c
SYS_SLIST_FOR_EACH_NODE(&ascs->ases, ase_node) {
```
like we do in `ase_find`

Thalley commented on subsys/bluetooth/audio/ascs_internal.h:
```diff
@@ -361,3 +361,6 @@ static inline const char *bt_ascs_reason_str(uint8_t reason)
 
 void ascs_ep_set_state(struct bt_audio_ep *ep, uint8_t state);
+
+int ascs_ase_config(struct bt_conn *conn, struct bt_audio_ep *ep,
+					struct bt_codec *codec);
```
```suggestion
int ascs_ase_config(struct bt_conn *conn, struct bt_audio_ep *ep,
		    struct bt_codec *codec);
```

Thalley reviewed (changes requested):
I think the approach is correct, but there's a couple of things that needs to be fixed. 

I would also strongly suggest to add an example of using the foreach and get info (preferably in the unicast server BSIM test (`tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_server_test.c`) to verify its correctness in CI

mkasenberg committed:
tests: bsim: audio: Add example usage of foreach ase

and get ase info to verify its correctness in CI.

Signed-off-by: Magdalena Kasenberg <magdalena.kasenberg@codecoup.pl>

mkasenberg commented:
After this refactor the stream has to be configured by the server app before calling the `bt_audio_unicast_server_config_ase()`. The function initializes the selected ASE and configures its codec. Finally, attaches the configured stream to the ASE.

The `bt_audio_unicast_server_foreach_ase()` allows to reach id, state and dir of each ASE, which the server app needs. This approach helped `bt_audio_ep` to stay safe and untouched by external code.

Added example usage of `bt_audio_unicast_server_foreach_ase()` into in the unicast server BSIM test.

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1784,6 +1784,58 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
+		bt_audio_ase_func_t func, void *user_data);
+
+/**
+ * @brief Init and configure ASE.
+ *
+ *  @param conn Connection object
+ *  @param ase_id ID of the ASE being configured
+ *  @param stream Configured stream object to be attached to the ASE
+ *  @param codec Codec configuration
+ *  @param qos_pref Audio Stream Quality of Service Preference
+ */
+int bt_audio_unicast_server_config_ase(struct bt_conn *conn, uint8_t ase_id,
```
Why do we need this, rather than using `bt_audio_stream_config`? 

Given the `ep_info`, we can now convert `ase_id` to `* ep`. 

One thing that does come to mind for me, is that the `qos_pref` would be missing, but we may get around by calling the `configure` callback with some way of indicating that this is a local call (which arguably can be done by comparing the `stream` or `ep` pointer.

Thalley replied:
The point is, I'd rather avoid adding a new function for this if we can, but you may have thought of some reason why we can't do that :)

mkasenberg replied:
I've consulted the issue with @MariuszSkamra today and we came to the conclusion that ase object does not exist until remote client triggers Codec Config.

So if the server needs to initiate the Codec Config before the client, an ase has to be created/inited first. Earlier I used `ase_get()` in `bt_unicast_server_foreach_ase()`:
```
	for (size_t i = 1; i <= ASE_COUNT; i++) {
		ase = ase_get(ascs, i);
```
 which silently created an ase for me. But after refactor of the loop:
```
	SYS_SLIST_FOR_EACH_NODE(&ascs->ases, ase_node) {
		ase = CONTAINER_OF(ase_node, struct bt_ascs_ase, node);
```
the `ascs->ases` list is empty, so does not work for me anymore.

So there are two options I can see:
- a function similar to `bt_audio_unicast_server_config_ase` which openly creates and configures the ase,
- or a public version of ase_get(), e.g. `struct bt_audio_ep * bt_audio_unicast_server_init_ase(struct bt_conn *conn, uint8_t ase_id);`, which openly creates and returns the ase, so we have the needed `ep` for `bt_audio_stream_config`.

Thalley replied:
`ase_get` is really a poor name for something that allocates a new ASE :( 

But yeah, I see the challenge. The `foreach` won't help us here, as the ASEs may not have been allocated for the ASCS instance. The modification to dynamically allocated ASEs made this a bit tricky :) 

Perhaps the best solution is to implement a `bt_audio_unicast_server_config_ase` as suggested. 

This should perhaps just take a `conn` and a `codec`, as the existing callback can be used to get the `stream` and `qos_pref`. 

The current suggested one can also work equally well if we omit the callback and then have `conn`, `stream`, `codec` and `qos_pref` as parameters. I'd like to avoid the `ase_id`, as the actual value of that is of no use to the application, and we have a few macros such as `ASE_DIR` and `ASE_UUID` that depend on ASE IDs being specific values. 

So I suggest to either have either
```c
int bt_audio_unicast_server_config_ase(struct bt_conn *conn, struct bt_codec *codec)
/* Depend on callback for the rest */
```

or 
```c
int bt_audio_unicast_server_config_ase(struct bt_conn *conn, struct bt_audio_stream *stream, struct bt_codec *codec, const struct bt_codec_qos_pref *qos_pref)
```

and then have the stack allocate the ASE ID.

mkasenberg replied:
I've selected the second version. Now the server is able to create all ASEs on a connection event callback, just before the PTS Upper Tester enables notifications, so PTS won't receive any too early. When PTS finally requests from IUT to initiate Codec Config, the server can find proper stream by ASE id from its stream array and pass the stream to the
```C
int bt_audio_unicast_server_config_ase(struct bt_conn *conn, struct bt_audio_stream *stream, struct bt_codec *codec, const struct bt_codec_qos_pref *qos_pref);
```

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1377,63 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+	struct bt_ascs_ase *ase;
+	struct bt_audio_ep *ep;
+
+	CHECKIF(conn == NULL || stream == NULL || codec == NULL) {
```
Also add `qos_pref`

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1377,63 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+
+	ep->qos_pref = *qos_pref;
+
+	bt_audio_stream_attach(conn, stream, ep, &ep->codec);
```
What if `stream->ep` is already set?

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1446,6 +1502,34 @@ static ssize_t ascs_config(struct bt_ascs *ascs, struct net_buf_simple *buf)
+		if (ase == NULL)
+			return;
+
+		bt_audio_ep_get_info(&ase->ep, &info);
```
As I suggested in `audio.h`, we should just return the `ep` pointer, and let the caller decide if they want to get the `info`

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1784,6 +1784,58 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
+struct bt_audio_ep_info {
+	/** The ID of the ASE */
+	uint8_t id;
+	/** The state of the endpoint */
+	uint8_t state;
```
We should use `enum bt_audio_state` and move the definitions from `stream.h` to `audio.h`

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -33,6 +33,35 @@ extern "C" {
 
 #define BT_AUDIO_BROADCAST_ID_SIZE               3 /* octets */
 
+/** Life-span states of ASE. Used only by internal APIs
+ *  dealing with setting ASE to proper state depending on operational
+ *  context.
+ *
+ * The values are defined by the ASCS spec and shall not be changed.
+ */
```
```suggestion
/** Endpoint states */
```

The comment is no longer valid, and should be updated

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1784,6 +1813,61 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
+ * @brief Structure holding information of audio stream endpoint
+ */
+struct bt_audio_ep_info {
+	/** The ID of the ASE */
```
```suggestion
	/** The ID of the endpoint */
```

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1784,6 +1813,61 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
  */
 int bt_audio_unicast_server_unregister_cb(const struct bt_audio_unicast_server_cb *cb);
 
+/**
+ * @brief Structure holding information of audio stream endpoint
+ */
```
```suggestion
/** Structure holding information of audio stream endpoint */
```

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1784,6 +1813,61 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
+	enum bt_audio_dir dir;
+};
+
+/**
+ * @brief Return structure holding information of audio stream endpoint
+ *
+ * @param ep   The audio stream endpoint object.
+ * @param info The structure object to be filled with the info.
```
```suggestion
/** @brief Return structure holding information of audio stream endpoint
 *
 *  @param ep   The audio stream endpoint object.
 *  @param info The structure object to be filled with the info.
```

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1784,6 +1813,61 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
+ */
+typedef void (*bt_audio_ep_func_t)(struct bt_audio_ep *ep, void *user_data);
+
+/**
+ * @brief Iterate through all endpoints of the given connection.
+ *
+ * @param conn Connection object
+ * @param func Function to call for each endpoint.
+ * @param user_data Data to pass to the callback function.
```
```suggestion
/** @brief Iterate through all endpoints of the given connection.
 *
 *  @param conn Connection object
 *  @param func Function to call for each endpoint.
 *  @param user_data Data to pass to the callback function.
```

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1784,6 +1813,61 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
+					bt_audio_ep_func_t func,
+					void *user_data);
+
+/**
+ * @brief Init and configure ASE.
```
```suggestion
/** @brief Initialize and configure a new ASE.
```

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1784,6 +1813,61 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
+ *  @param stream Configured stream object to be attached to the ASE
+ *  @param codec Codec configuration
+ *  @param qos_pref Audio Stream Quality of Service Preference
+ */
```
Missing `@return` or `@retval`

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1784,6 +1813,61 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
+ *
+ * @param ep   The audio stream endpoint object.
+ * @param info The structure object to be filled with the info.
+ */
```
Missing `@return` or `@retval`

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1378,68 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
 	return 0;
 }
 
+int bt_audio_unicast_server_config_ase(struct bt_conn *conn,
+		struct bt_audio_stream *stream, struct bt_codec *codec,
+		const struct bt_codec_qos_pref *qos_pref)
```
Indent

```suggestion
int bt_audio_unicast_server_config_ase(struct bt_conn *conn, struct bt_audio_stream *stream, 
				       struct bt_codec *codec, 
				       const struct bt_codec_qos_pref *qos_pref)
```

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1378,68 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+	struct bt_ascs_ase *ase;
+	struct bt_audio_ep *ep;
+
+	CHECKIF(conn == NULL || stream == NULL || codec == NULL ||
+		qos_pref == NULL) {
```
```suggestion
	CHECKIF(conn == NULL || stream == NULL || codec == NULL || qos_pref == NULL) {
```

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1378,68 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+	}
+
+	ascs = ascs_get(conn);
+	ep = stream->ep;
+
+	if (ep == NULL) {
```
`stream->ep` shall always be NULL in this case. If the stream is already configured to a different endpoint, it should be released first. Suggest to return error if `stream->ep` is not NULL

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1378,68 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+		ep = &ase->ep;
+	}
+
+	switch (ep->status.state) {
+	/* Valid only if ASE_State field = 0x00 (Idle) */
+	case BT_AUDIO_EP_STATE_IDLE:
+	 /* or 0x01 (Codec Configured) */
+	case BT_AUDIO_EP_STATE_CODEC_CONFIGURED:
+	 /* or 0x02 (QoS Configured) */
+	case BT_AUDIO_EP_STATE_QOS_CONFIGURED:
+		break;
+	default:
+		LOG_ERR("Invalid state: %s", bt_audio_ep_state_str(
+			ep->status.state));
+		return -EBADMSG;
+	}
```
`bt_audio_unicast_server_config_ase` shall configure a new ASE. If the ASE is any other state than idle, then we should return an error.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1378,68 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+
+	if (ep == NULL) {
+		sys_snode_t *ase_node;
+		uint8_t ase_next_id = 1;
+
+		SYS_SLIST_FOR_EACH_NODE(&ascs->ases, ase_node) {
+			ase_next_id++;
+		}
+
+		ase = ase_get(ascs, ase_next_id);
+		if (ase == NULL) {
+			LOG_WRN("No free ASE found for config ASE ID 0x%02x", ase_next_id);
+			return -ENOTSUP;
+		}
```
This function shall configure a new ASE. 

If there already is an ASE with ID 1, then `ase_get` returns that one, which is not what we want. Instead what we should do is call `ase_find` until that returns NULL, which indicates an unused ASE. Once we then have the "free index", we can call `ase_new` with that to get a new unused ASE.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1446,6 +1508,33 @@ static ssize_t ascs_config(struct bt_ascs *ascs, struct net_buf_simple *buf)
+	return 0;
+}
+
+void bt_audio_unicast_server_foreach_ep(struct bt_conn *conn,
+		bt_audio_ep_func_t func, void *user_data)
```
```suggestion
void bt_audio_unicast_server_foreach_ep(struct bt_conn *conn, bt_audio_ep_func_t func, 
					void *user_data)
```

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1446,6 +1508,33 @@ static ssize_t ascs_config(struct bt_ascs *ascs, struct net_buf_simple *buf)
+	return 0;
+}
+
+void bt_audio_unicast_server_foreach_ep(struct bt_conn *conn,
```
This function should probably be put in `unicast_server.c` instead.

mkasenberg replied:
In `bt_audio_unicast_server_foreach_ep()` there is usage of static `ascs_get()`. Can we leave the function here or is it better to create non-static version of `ascs_get()`?

Thalley replied:
I would suggest to move `ascs_get()` to `ascs_internal.h` and rename to `bt_ascs_get`

mkasenberg replied:
I've faced a problem with circular imports when moving all contents of `bt_audio_unicast_server_foreach_ep` and `bt_audio_unicast_server_config_ase` into `unicast_server.c`. To fix this, the `struct bt_ascs_ase` and `struct bt_ascs` would have to be moved into `endpoint.h`. Finally I would need `bt_ascs_get`, `bt_ascs_find` and `bt_ascs_new` in `ascs_internal.h`...

So to avoid major changes, in `ascs_internal.h` there are now:
```C
int bt_ascs_config_ase(struct bt_conn *conn, struct bt_audio_stream *stream, struct bt_codec *codec,
		       const struct bt_codec_qos_pref *qos_pref);

void bt_ascs_foreach_ep(struct bt_conn *conn, bt_audio_ep_func_t func, void *user_data);
```
which are called from `bt_audio_unicast_server_config_ase` and `bt_audio_unicast_server_foreach_ep` in `unicast_server.c`. That way we would keep the naming convention and leave room for the future refactors

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1446,6 +1508,33 @@ static ssize_t ascs_config(struct bt_ascs *ascs, struct net_buf_simple *buf)
 	return buf->size;
 }
 
+int bt_audio_ep_get_info(const struct bt_audio_ep *ep, struct bt_audio_ep_info *info)
+{
+	info->id = ep->status.id;
+	info->state = ep->status.state;
+	info->dir = ep->dir;
+
+	return 0;
+}
```
This should be put in `stream.c` I think, so that it can be used as the unicast client as well

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1446,6 +1508,33 @@ static ssize_t ascs_config(struct bt_ascs *ascs, struct net_buf_simple *buf)
+	SYS_SLIST_FOR_EACH_NODE(&ascs->ases, ase_node) {
+		ase = CONTAINER_OF(ase_node, struct bt_ascs_ase, node);
+
+		if (ase == NULL) {
+			return;
+		}
+
```
```suggestion
```

`CONTAINER_OF` doesn't return NULL

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1446,6 +1508,33 @@ static ssize_t ascs_config(struct bt_ascs *ascs, struct net_buf_simple *buf)
+void bt_audio_unicast_server_foreach_ep(struct bt_conn *conn,
+		bt_audio_ep_func_t func, void *user_data)
+{
+	struct bt_ascs_ase *ase;
+	sys_snode_t *ase_node;
+	struct bt_ascs *ascs = ascs_get(conn);
+
+	SYS_SLIST_FOR_EACH_NODE(&ascs->ases, ase_node) {
+		ase = CONTAINER_OF(ase_node, struct bt_ascs_ase, node);
```
```suggestion
	sys_snode_t *ase_node;
	struct bt_ascs *ascs = ascs_get(conn);

	SYS_SLIST_FOR_EACH_NODE(&ascs->ases, ase_node) {
		struct bt_ascs_ase *ase = CONTAINER_OF(ase_node, struct bt_ascs_ase, node);
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_server_test.c:
```diff
@@ -34,6 +34,15 @@ static const struct bt_data unicast_server_ad[] = {
 CREATE_FLAG(flag_connected);
 CREATE_FLAG(flag_stream_configured);
 
+static void print_ase_info(struct bt_audio_ep *ep, void *user_data)
+{
+	struct bt_audio_ep_info info;
+
+	bt_audio_ep_get_info(ep, &info);
+	printk("ASE info: id %u state %u dir %u\n", info.id, info.state,
+	       info.dir);
+}
```
Suggest to also add this to the shell `subsys/bluetooth/shell/audio.c` with a command to print ASE info

Thalley reviewed (changes requested):
Please split the endpoint get info and the unicast server configure into two commits, as they are not depending on each other.

mkasenberg committed:
LE Audio: Add helper functions to get endpoints info

The functions allow to return structure holding information of audio
stream endpoint.

Signed-off-by: Magdalena Kasenberg <magdalena.kasenberg@codecoup.pl>

mkasenberg committed:
shell: audio: Add print ASE info command

The command prints id, state and dir of all ASEs of default connection.

Signed-off-by: Magdalena Kasenberg <magdalena.kasenberg@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1378,62 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
 	return 0;
 }
 
+int bt_audio_unicast_server_config_ase(struct bt_conn *conn, struct bt_audio_stream *stream,
+				       struct bt_codec *codec,
+				       const struct bt_codec_qos_pref *qos_pref)
```
Ideally this would also, based on the name, be moved to `unicast_server.c`

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1378,62 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+		return -EALREADY;
+	}
+
+	for (int i = 1; i <= ASE_COUNT; i++) {
+		ase = ase_find(ascs, i);
+
+		if (ase == NULL) {
+			ase = ase_new(ascs, i);
+			break;
+		}
+	}
```
Can you please add a small comment stating why we are doing this? Just a small explanation stating that this is to get a "free" ASE

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1378,6 +1378,62 @@ static int ase_config(struct bt_ascs *ascs, struct bt_ascs_ase *ase,
+		return -EBADMSG;
+	}
+
+	if (ascs_ep_set_codec(ep, codec->id, sys_le16_to_cpu(codec->cid),
+			      sys_le16_to_cpu(codec->vid), NULL, 0, &ep->codec)) {
+		return -ENOTSUP;
+	}
```
Suggest to return the error value of `ascs_ep_set_codec` instead

Thalley reviewed (changes requested):
A few minor comments, otherwise it looks good.

Thanks for your patience with my reviews and effort to update it this quickly :)
